Implemented custom data types support in DBMSs

SilverFire opened the pull request:
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #7640, #15377

This is the alternative implementation of PR #14819, #14836

TODO:

- [x] Code idea
- [x] Discuss and approve it @yiisoft/reviewers @yiisoft/core-developers 
- [x] Add popular types for PgSQL, MySQL
- [x] Enhance tests coverage
- [x] Write docs
- [x] Merge

SilverFire committed:
Implemented custom data types support in DBMSs

replaces PR #14836, #14819

SilverFire commented:
In comparison to previous implementations, this one provides DBMS-agnostic classes of expressions (`ArrayExpression`, `JsonExpression`) that are simply a DTO.

Expression building is moved builders that implement `ExpressionBuilderInterface` and are DBMS-specific. I would like to implement wider JSON support in DB abstraction layer based on this PR.

Dear reviewers, please ignore imperfect PHPDocs. They are, but we need to agree design first. Thank you

samdark commented on framework/db/ExpressionBuilder.php:
```diff
@@ -0,0 +1,24 @@
+    public function build(ExpressionInterface $expression, &$params = [])
+    {
+        $params = array_merge($params, $expression->params);
+        return $expression->__toString();
```
I'd explicitly add a method to the interface instead of using magical `__toString`.

SilverFire replied:
This expression builder is specific for `yii\db\Expression`. Which kind of method would you add to the interface?

samdark replied:
Nevermind. It's fine.

samdark commented on framework/db/pgsql/JsonExpressionBuilder.php:
```diff
@@ -0,0 +1,64 @@
+use yii\db\Query;
+
+/**
+ * Class JsonExpressionBuilder builds [[JsonExpression]] for PostgreSQL DBMS.
```
We'll be able to add another builder for MySQL, correct?

SilverFire replied:
Right

samdark commented:
Love the approach! I'm all for it.

zuozp8 commented on framework/db/ExpressionBuilderTrait.php:
```diff
@@ -0,0 +1,28 @@
+ * @author Dmytro Naumenko <d.naumenko.a@gmail.com>
+ * @since 2.0.14
+ */
+trait ExpressionBuilderTrait
```
i see no usecase where you need splitted interface and trait, i would merge ExpressionBuilderTrait and ExpressionBuilderInterface into abstract class

SilverFire replied:
Disagreed. I can build my own extending tree and implement class wherever I want.
Abstract class forces me to add it in extending tree root and makes reusing harder.

zuozp8 replied:
you're right, separation will be needed for example in https://github.com/yiisoft/yii2/pull/15348/files#r156621010

zuozp8 commented on framework/db/QueryBuilder.php:
```diff
@@ -145,6 +166,43 @@ public function build($query, $params = [])
+    {
+        $className = get_class($expression);
+        if (!isset($this->expressionBuilders[$className])) {
+            foreach ($this->expressionBuilders as $expressionClass => $builderClass) {
```
i would rename `$builderClass` into `$builderClassOrObject`

zuozp8 commented on framework/db/QueryBuilder.php:
```diff
@@ -145,6 +166,43 @@ public function build($query, $params = [])
+        if (!isset($this->expressionBuilders[$className])) {
+            foreach ($this->expressionBuilders as $expressionClass => $builderClass) {
+                if (is_subclass_of($expression, $expressionClass)) {
+                    $this->expressionBuilders[$className] = $builderClass;
```
it's a bit counter intuitive that behavior depends on the order of execution
imagine:
we have `$this->expressionBuilders = ['…\ParentExpression' => '…\ConcreteBuilder']
we have `class ChildExpression extends …\ParentExpression`

if we call `->buildExpression($childExpression)` and then `->buildExpression($parentExpression)` then we create separate builder instances for each class

if we call `->buildExpression($parentExpression)` and then `->buildExpression($childExpression)` then we create one instance that is shared between classes

zuozp8 replied:
if we have
`$this->expressionBuilders = ['…\GrandParentExpression' => '…\Builder1', '…\ParentExpression' => '…\Builder2']`
and we pass instance of `…\ChildExpression` to be built, then we would use Builder1, but it seems Builder 2 would be more suitable

it may be problem as setExpressionBuilders is only appending to the list, it's worth documenting at least

SilverFire replied:
Yes, that my be confusing. Of course we can go deeper and find the most suitable builder with `class_parents` and `class_impements`, but it will affect performance and increase code complexity. Do we need it?

I'm OK with documenting this behavior.

hiqsol replied:
Considering `array_merge` appends builders to the end of list I think builders should be processed in reverse order which allows to solve the problem.

zuozp8 commented on framework/db/ExpressionInterface.php:
```diff
@@ -0,0 +1,19 @@
+ * @author Dmytro Naumenko <d.naumenko.a@gmail.com>
+ * @since 2.0.14
+ */
+interface ExpressionInterface
```
how about calling it QueryFragmentInterface and make Query implement in the future?

It would make us able to use subqueries everywhere (everywhere where we have expression support) and simplify the code. Otherwise we'll never support subqueries in non-obvious places like inside `order by`

SilverFire replied:
Not sure about naming, but in general it's a good idea originally suggested by @hiqsol [in previous PR](https://github.com/yiisoft/yii2/pull/14836#issuecomment-330166721)

SilverFire replied:
Will try to adopt it in next changes iteration

klimov-paul commented:
I do not like the idea - it seems to overcomplicated for the purpose.
If you wish to grad develoepr ability to configure particular condition building it is better to be done via [QueryBuilder::$conditionBuilders](https://github.com/yiisoft/yii2/blob/master/framework/db/QueryBuilder.php#L50-LL68). All we need to be is add an ability to specify particular builder as a PHP callback (or a object matching certain interface if you can not live without it) with following signature:

```php
function (QueryBuilder $queryBuilder, string $operator, array $operands, array &$params) : string {
    return 'SQL';
}
```
Particular condition builder may accept different data types for the `$operands` including `db\Expression` or any other complex object.

Obviously `QueryBuilder` will need an additional public setter for the `$conditionBuilders`, like following:

```php
public function setConditionBuilders(array $conditionBuilders)
{
    $this->conditionBuilders = array_merge($this->conditionBuilders, $conditionBuilders);
}
```

Such approach will be much more flexible, allowing of handling both custom data types and operators.

> Implemented custom data types support in DBMSs

Such statement is incorrect regardless to this PR, as well @SilverFire's late attempts of solving original issue, since it covers only query condition composition, while Array DB type support require handling of the array data at the `INSERT` and `UPDATE` statement as well.
In order to grand support for `ARRAY` or `JSON` DB data types, there should be granted ability to pass  PHP arrays as a column value for the `insert()` and `update()` functions.
If you wish to solve the issue using custom TYPE-based classes - they should be implemented at the level of `yii\db\ColumnSchema` instead of  database expressions.

klimov-paul commented:
Having both `QueryBuilder::$conditionBuilders` and `QueryBuilder::$expressionBuilders` intoduces 2 different appraoches fot doing the same thing. If it is an object interface for condition composition you want - `QueryBuilder::$conditionBuilders` should be entirely removed in favor of `QueryBuilder::$expressionBuilders` and any complex operators should be composed via custom `db\Expressions` classes.
E.g. instead of following code:

```php
$rows = (new Query())
    ->from('item')
    ->where(['in', 'id', [1, 2, 3]])
    ->andWhere(['between', 'date', '2017-09-01', '2017-10-01'])
    ->all();
```

the following code should be used:

```php
$rows = (new Query())
    ->from('item')
    ->where(new InExpression('id', [1, 2, 4]))
    ->adnWhere(new BetweenExpression('date', '2017-09-01', '2017-10-01'))
    ->all();
```

SilverFire commented:
Generally you suggestion it makes sense, but in wise-versa logical direction.

Expression – is a Data Transfer Object (DTO) for dataset, that can be somehow compiled to some SQL statement (operators, string, array, JSON, etc)
Condition – is an aggregate of multiple Expressions that can be compiled to SQL condition.

In means that we can create create condition classes (`SimpleCondition`, `InCondition`, `ExistsCondition`) that implement `ExpressionInterface` and create builders for each of them (or for logical groups of them). This kind of refactoring should also reduce complexity of `QueryBuilder` and make each expression more testable.

SilverFire commented:
Ha, you've posted nearly the same while I composed my message.

> E.g. instead of following code:
> ```php
> $rows = (new Query())
>     ->from('item')
>     ->where(['in', 'id', [1, 2, 3]])
>     ->andWhere(['between', 'date', '2017-09-01', '2017-10-01'])
>     ->all();
> ```
>
> the following code should be used:
> 
> ```php
> $rows = (new Query())
>     ->from('item')
>     ->where(new InExpression('id', [1, 2, 4]))
>     ->adnWhere(new BetweenExpression('date', '2017-09-01', '2017-10-01'))
>     ->all();
> ```

In order to keep BC we can transform the first into the second automatically.

sergeymakinen commented:
In general, @klimov-paul already expressed my initial suggestions for the PR.

I'll review it in detail a bit later but I love the idea and I really like the way it's implemented. @SilverFire, keep it rolling! 👍

SilverFire committed:
Added JsonExpressionBuilder for MySQL, added tests, updated docs

SilverFire committed:
Intermediate

SilverFire committed:
Added JSON and array support to AR layer, added PdoValue

SilverFire commented on framework/base/Controller.php:
```diff
@@ -5,6 +5,8 @@
  * @license http://www.yiiframework.com/license/
  */
 
+// TODO titti
```
Nevermind :)

SilverFire commented:
Added another bunch of effort in this PR.

- Implemented JSON and Array support for PgSQL (in ActiveRecord layer as well)
- Added `PdoValue` class to deprecate array-based (`[$value, \PDO::PARAM_LOG]`) type casting

Future iterations plan: 

- JSON for MySQL
- Merge `expressionBuilders` and `conditionBuilders` as discussed above with @klimov-paul 

Sorry for long delays between commits. AR & DBAL is one of the most complex parts in Yii so it's not easy to apply changes there and stay concentrated for a long time.

samdark commented on docs/guide/db-query-builder.md:
```diff
@@ -758,3 +758,15 @@ $unbufferedDb->close();
+
+### PostgreSQL arrays
+
+TBD
```
Should not be merged w/o it done.

samdark commented on docs/guide/db-query-builder.md:
```diff
@@ -758,3 +758,15 @@ $unbufferedDb->close();
+
+### PostgreSQL `ILIKE`
+
+TBD
```
Should not be merged w/o it done.

samdark commented on framework/UPGRADE.md:
```diff
@@ -56,6 +56,11 @@ Upgrade from Yii 2.0.13
+  `instanceof yii\db\Expression` in your code, you might consider changing that to checking for the interface and use the newly
+  introduced methods to retrieve the expression content.
+  
+* TODO: `['value', \PDO::PARAM_STR]` syntax deprecation.
```
Should not be merged w/o it done.

samdark commented on framework/db/ArrayExpression.php:
```diff
@@ -0,0 +1,83 @@
+/**
+ * Class ArrayExpression represents an array SQL expression.
+ *
+ * Expressions of this type can be used for example in conditions, like:
```
Delete "for example".

samdark commented on framework/db/ArrayExpression.php:
```diff
@@ -0,0 +1,83 @@
+     */
+    protected $value;
+    /**
+     * @var int
```
phpdoc needed.

samdark commented on framework/db/ArrayExpression.php:
```diff
@@ -0,0 +1,83 @@
+     * @param string|null $type the type of the array elements. Defaults to `null` which means the type is
+     * not explicitly specified. In case when type is not specified explicitly and DBMS can not guess it from the context,
+     * SQL error will be raised.
+     * @param int $dimension
```
phpdoc needed.

samdark commented on framework/db/ColumnSchema.php:
```diff
@@ -114,12 +115,18 @@ public function dbTypecast($value)
+        if (is_array($value) && count($value) === 2 && isset($value[1]) && in_array($value[1], $this->getPdoParamTypes(), true)) {
+            return new PdoValue($value[0], $value[1]);
+            // Backward compatibility thing, but too unreliable. Array [1,2] will pass into this condition and transform
+            // into PdoValue class. However, arrays did not come this method earlier so should not be a big deal. TODO: consider.
```
> However, arrays did not come this method earlier

Are you sure about that? Was it impossible to supply an array here?

SilverFire replied:
It was used in `dbTypecast()` and `phpTypecast()` methods only.

Here are the complete list of usages:

**dbTypecast()**

 - `QueryBuilder::insert()` prefixed with `!is_array()`
https://github.com/yiisoft/yii2/blob/2ba98c6d5befb667fa4c476951f01a53fee07e02/framework/db/QueryBuilder.php#L195-L197

- `QueryBuilder::batchInsert()` wrapped in `!is_array()` condition
https://github.com/yiisoft/yii2/blob/2ba98c6d5befb667fa4c476951f01a53fee07e02/framework/db/QueryBuilder.php#L279-L281

- `QueryBuilder::update()` prefixed with `!is_array()`
https://github.com/yiisoft/yii2/blob/2ba98c6d5befb667fa4c476951f01a53fee07e02/framework/db/QueryBuilder.php#L344-L346

- MySQL `QueryBuilder::insert()` prefixed with `!is_array()`
https://github.com/yiisoft/yii2/blob/2ba98c6d5befb667fa4c476951f01a53fee07e02/framework/db/mysql/QueryBuilder.php#L259-L251

- Oracle `QueryBuilder::insert()` prefixed with `!is_array()`
https://github.com/yiisoft/yii2/blob/2ba98c6d5befb667fa4c476951f01a53fee07e02/framework/db/oci/QueryBuilder.php#L212-L215

- Oracle `QueryBuilder::batchInsert()` wrapped in `!is_array()` condition
https://github.com/yiisoft/yii2/blob/2ba98c6d5befb667fa4c476951f01a53fee07e02/framework/db/oci/QueryBuilder.php#L268

- PgSQL `QueryBuilder::batchInsert()` wrapped in `!is_array()` condition
https://github.com/yiisoft/yii2/blob/master/framework/db/pgsql/QueryBuilder.php#L305-L307

- SQLite `QueryBuilder::batchInsert()` wrapped in `!is_array()` condition
https://github.com/yiisoft/yii2/blob/master/framework/db/sqlite/QueryBuilder.php#L101-L103

**phpTypecast()**

- `ActiveRecord::populateRecord()` – potentially dangerous
https://github.com/yiisoft/yii2/blob/5a8c3d537ba14dbff02b2adee9072d381b9eb6b7/framework/db/ActiveRecord.php#L418-L427

- `ActiveRecord::insertInternal()` – have you ever seen primary key represented as array of arrays?
https://github.com/yiisoft/yii2/blob/5a8c3d537ba14dbff02b2adee9072d381b9eb6b7/framework/db/ActiveRecord.php#L515-L517

- Also used to typecast default values: [1](https://github.com/yiisoft/yii2/blob/1501c659aca7c50fa5e4f6fbe5ca549acc946c86/framework/db/cubrid/Schema.php#L335), [2](https://github.com/yiisoft/yii2/blob/4e176ca3259083a8c441c70182bc10e96cbfec12/framework/db/mssql/Schema.php#L386), [3](https://github.com/yiisoft/yii2/blob/f9fd4da8de6c0272cd8956e94da1656b41f1416c/framework/db/mysql/Schema.php#L303), [4](https://github.com/yiisoft/yii2/blob/185209957e11ba6a427c1aa418dfcba7734e8b7b/framework/db/oci/Schema.php#L416), [5](https://github.com/yiisoft/yii2/blob/284dae6a55e44590d495831b9761f1ca50a8bc76/framework/db/pgsql/Schema.php#L565), [6](https://github.com/yiisoft/yii2/blob/2d672b67223b8d930f0f444b113d0a13fbd551c0/framework/db/sqlite/Schema.php#L364), but should not be harmful.

Considering that tests are not broken, I tend to say that the change is not a serious breaking.

samdark replied:
Yes. There could be composite PKs. These contain multiple columns.

SilverFire replied:
No, composite PK is array of strings: `['id', 'version']`, not array of arrays `[['id'], ['version']]`

samdark replied:
Then it should be fine.

samdark commented on framework/db/ColumnSchema.php:
```diff
@@ -143,4 +150,9 @@ protected function typecast($value)
         return $value;
     }
+
+    private function getPdoParamTypes()
```
phpdoc

samdark commented on framework/db/Command.php:
```diff
@@ -337,8 +337,8 @@ public function bindValue($name, $value, $dataType = null)
      * e.g. `[':name' => 'John', ':age' => 25]`. By default, the PDO type of each value is determined
-     * by its PHP type. You may explicitly specify the PDO type by using an array: `[value, type]`,
-     * e.g. `[':name' => 'John', ':profile' => [$profile, \PDO::PARAM_LOB]]`.
+     * by its PHP type. You may explicitly specify the PDO type by using a PdoValue class: `new PdoValue(value, type)`,
```
Full namespace needed for PdoValue and `[[   ]]`.

samdark commented on framework/db/PdoValue.php:
```diff
@@ -0,0 +1,51 @@
+ * Class PdoValue represents a $value that should be bound to PDO with defined $type and
+ * should not be transformed before binding.
+ *
+ * TODO: usage example description, update docs in framework to follow this change.
```
Should be done before merge.

samdark commented on framework/db/mysql/JsonExpressionBuilder.php:
```diff
@@ -0,0 +1,42 @@
+        $value = $expression->getValue();
+
+        if ($value instanceof Query) {
+            list ($sql, $params) = $this->queryBuilder->build($value, $params);
```
Is it OK that params aren't used?

SilverFire replied:
They are passed be reference

samdark commented on framework/db/pgsql/ArrayExpressionBuilder.php:
```diff
@@ -0,0 +1,134 @@
+
+/**
+ * Class ArrayExpressionBuilder builds [[ArrayExpression]] for PostgreSQL DBMS.
+ * TODO: tests
```
Should be done before merging.

samdark commented on framework/db/pgsql/ArrayParser.php:
```diff
@@ -0,0 +1,86 @@
+namespace yii\db\pgsql;
+
+/**
+ * TODO: phpdoc, tests
```
Should be done before merging.

samdark commented on framework/db/pgsql/ColumnSchema.php:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    /**
+     * TODO: phpdoc
```
Should be done before merging.

samdark commented on framework/db/pgsql/JsonExpressionBuilder.php:
```diff
@@ -0,0 +1,55 @@
+
+        if ($value instanceof Query) {
+            list ($sql, $params) = $this->queryBuilder->build($value, $params);
+            return "($sql)" . $this->getTypecast($expression);
```
It is good that params aren't used?

SilverFire replied:
This variable is passed by reference, so it is OK

samdark commented on framework/db/ExpressionBuilderInterface.php:
```diff
@@ -0,0 +1,23 @@
+     * @param array $params the binding parameters.
+     * @return string the raw SQL that will not be additionally escaped or quoted.
+     */
+    public function build(ExpressionInterface $expression, &$params = []);
```
Does it make sense to declare params as `array &$params = []`?

SilverFire replied:
Not sure

samdark replied:
I think it does.

samdark commented on framework/db/PdoValue.php:
```diff
@@ -0,0 +1,51 @@
+        $this->type = $type;
+    }
+
+    public function getValue()
```
phpdoc

samdark commented on framework/db/PdoValue.php:
```diff
@@ -0,0 +1,51 @@
+        return $this->value;
+    }
+
+    public function getType()
```
phpdoc

samdark commented on framework/db/QueryBuilder.php:
```diff
@@ -1584,12 +1626,8 @@ public function buildSimpleCondition($operator, $operands, &$params)
-            }
-
-            return "$column $operator {$value->expression}";
+        } elseif ($value instanceof ExpressionInterface) {
```
Could be just `if`.

samdark commented on framework/db/QueryBuilder.php:
```diff
@@ -1584,12 +1626,8 @@ public function buildSimpleCondition($operator, $operands, &$params)
-            return "$column $operator {$value->expression}";
+        } elseif ($value instanceof ExpressionInterface) {
+            return "$column $operator {$this->buildExpression($value, $params)}";
         } elseif ($value instanceof Query) {
```
Could be just `if`.

samdark commented on framework/db/pgsql/ArrayParser.php:
```diff
@@ -0,0 +1,86 @@
+     */
+    public function parse($value)
+    {
+        if (is_null($value)) {
```
`$value === null`

samdark commented on framework/db/pgsql/ArrayParser.php:
```diff
@@ -0,0 +1,86 @@
+        return $this->parseArray($value);
+    }
+
+    private function parseArray($value, &$i = 0)
```
phpdoc

samdark commented on framework/db/pgsql/ArrayParser.php:
```diff
@@ -0,0 +1,86 @@
+    private function parseArray($value, &$i = 0)
+    {
+        $result = [];
+        for(++$i; $i < strlen($value); ++$i) {
```
1. `$value` could be in UTF so `strlen` won't work.
2. Formatting.
3. `strlen` should not be called in condition part of `for`. Either in pre-condition or out of cycle.

samdark commented on framework/db/pgsql/ArrayParser.php:
```diff
@@ -0,0 +1,86 @@
+                    if (empty($result)) { // `{}` case
+                        $result[] = null;
+                    }
+                    if (in_array($value[$i + 1], [$this->delimiter, '}'])) { // `{,}` case
```
Specify `true` as 3rd argument.

samdark commented on framework/db/pgsql/ArrayParser.php:
```diff
@@ -0,0 +1,86 @@
+        return $result;
+    }
+
+    private function parseString($value, &$i)
```
phpdoc

samdark commented on framework/db/pgsql/ArrayParser.php:
```diff
@@ -0,0 +1,86 @@
+        $isQuoted = $value[$i] === '"';
+        $stringEndChars = $isQuoted ? ['"'] : [$this->delimiter, '}'];
+        $result = '';
+        for ($i += $isQuoted ? 1 : 0; $i < strlen($value); ++$i) {
```
1. `$value` could be in UTF so `strlen` won't work.
2. `strlen` should not be called in condition part of `for`. Either in pre-condition or out of cycle.

SilverFire replied:
1. Added test to prove it works OK. Can you suggest an example, when it does not? https://github.com/yiisoft/yii2/pull/15348/commits/8f4c56912f90da3a6b5dbe8faeb00ab5de48faff
2. Done

samdark commented on framework/db/pgsql/ArrayParser.php:
```diff
@@ -0,0 +1,86 @@
+        for ($i += $isQuoted ? 1 : 0; $i < strlen($value); ++$i) {
+            if (in_array($value[$i], ['\\', '"'], true) && in_array($value[$i + 1], [$value[$i], '"'], true)) {
+                ++$i;
+            } elseif (in_array($value[$i], $stringEndChars)) {
```
Specify `true` as 3rd argument.

samdark commented on framework/db/pgsql/ColumnSchema.php:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    /**
+     * TODO: phpdoc
```
Should be done before merging.

SilverFire committed:
Enhanced docs, added FakeExpressionBuilder

SilverFire committed:
Added test to make sure ArrayParser works with UTF8

SilverFire committed:
Added PHPDocs

SilverFire committed:
Trying to switch to PgSQL 9.6 on Travis

SilverFire committed:
Added ConditionInterface and some implementation, added PdoValueBuilder

SilverFire committed:
Added rest of conditions. Cleared QueryBuilder out of old code

SilverFire committed:
Returned QueryBuilder::build*Conditon for BC

SilverFire committed:
Updated all QueryBuilders to follow API changes

SilverFire committed:
Merge branch 'master' into pgsql-arrays-expression-v3

SilverFire committed:
Fixed pgsql dump

samdark commented on framework/db/conditions/BetweenCondition.php:
```diff
@@ -0,0 +1,92 @@
+            throw new InvalidParamException("Operator '$operator' requires three operands.");
+        }
+
+        return (new static($operands[0], $operator, $operands[1], $operands[2]));
```
No need for extra `(` and `)` wrapping whole `new` clause.

samdark commented on framework/db/conditions/BetweenConditionBuilder.php:
```diff
@@ -0,0 +1,58 @@
+use yii\db\ExpressionBuilderInterface;
+use yii\db\ExpressionBuilderTrait;
+use yii\db\ExpressionInterface;
+use yii\db\Query;
```
Not used. Remove.

samdark commented on framework/db/conditions/ConditionInterface.php:
```diff
@@ -0,0 +1,28 @@
+<?php
```
Standard Yii header missing.

SilverFire replied:
It will be added during release process automatically

samdark commented on framework/db/conditions/BetweenConditionBuilder.php:
```diff
@@ -0,0 +1,58 @@
+<?php
```
Standard Yii header missing.

SilverFire replied:
It will be added during release process automatically

samdark commented on framework/db/conditions/BetweenCondition.php:
```diff
@@ -0,0 +1,92 @@
+<?php
+
```
Standard Yii header missing.

SilverFire replied:
It will be added during release process automatically

samdark commented on framework/db/conditions/AndCondition.php:
```diff
@@ -0,0 +1,22 @@
+<?php
+
```
Standard Yii header missing.

SilverFire replied:
It will be added during release process automatically

samdark commented on framework/db/conditions/ConjunctionCondition.php:
```diff
@@ -0,0 +1,47 @@
+     */
+    public static function fromArrayDefinition($operator, $operands)
+    {
+        return (new static($operands));
```
No need for `(` and `)`.

samdark commented on framework/db/conditions/ExistsCondition.php:
```diff
@@ -0,0 +1,64 @@
+            throw new InvalidParamException('Subquery for EXISTS operator must be a Query object.');
+        }
+
+        return (new static($operator, $operands[0]));
```
No need for `(` and `)`.

samdark commented on framework/db/conditions/HashCondition.php:
```diff
@@ -0,0 +1,45 @@
+
+namespace yii\db\conditions;
+
+use yii\base\InvalidParamException;
```
Not used. Remove.

samdark commented on framework/db/conditions/InCondition.php:
```diff
@@ -0,0 +1,84 @@
+            throw new InvalidParamException("Operator '$operator' requires two operands.");
+        }
+
+        return (new static($operands[0], $operator, $operands[1]));
```
`(` and `)` aren't needed.

samdark commented on framework/db/conditions/InConditionBuilder.php:
```diff
@@ -0,0 +1,165 @@
+
+        if ($column instanceof \Traversable || ((is_array($column) || $column instanceof \Countable) && count($column) > 1)) {
+            return $this->buildCompositeInCondition($operator, $column, $values, $params);
+        } elseif (is_array($column)) {
```
No need for else since previously there was a return.

SilverFire replied:
It will be added during release process automatically

samdark commented on framework/db/conditions/LikeCondition.php:
```diff
@@ -0,0 +1,71 @@
+     *   If it is an empty array the generated expression will  be a `false` value if operator is `LIKE` or `OR LIKE`
+     *   and empty if operator is `NOT LIKE` or `OR NOT LIKE`.
+     */
+    public function __construct($column, $operator, $value)
```
Is it for phpdoc only?

SilverFire replied:
Yes

samdark commented on framework/db/conditions/LikeCondition.php:
```diff
@@ -0,0 +1,71 @@
+            throw new InvalidParamException("Operator '$operator' requires two operands.");
+        }
+
+        $condition = (new static($operands[0], $operator, $operands[1]));
```
No need for `(` and `)`.

samdark commented on framework/db/conditions/LikeConditionBuilder.php:
```diff
@@ -0,0 +1,96 @@
+                $phName = $this->queryBuilder->bindParam(empty($escape) ? $value : ('%' . strtr($value, $escape) . '%'), $params);
+            }
+            $escapeSql = '';
+            if ($this->escapeCharacter !== null) {
```
Could be moved out of foreach.

samdark commented on framework/db/conditions/NotCondition.php:
```diff
@@ -0,0 +1,50 @@
+            throw new InvalidParamException("Operator '$operator' requires exactly one operand.");
+        }
+
+        return (new static(array_shift($operands)));
```
`(` and `)` aren't needed.

samdark commented on framework/db/conditions/SimpleCondition.php:
```diff
@@ -0,0 +1,78 @@
+            throw new InvalidParamException("Operator '$operator' requires two operands.");
+        }
+
+        return (new static($operands[0], $operator, $operands[1]));
```
No need for `(` and `)`.

samdark commented on framework/db/cubrid/QueryBuilder.php:
```diff
@@ -46,18 +46,14 @@ class QueryBuilder extends \yii\db\QueryBuilder
-        '!' => '!!',
-    ];
-
+    protected function defaultExpressionBuilders()
```
phpdoc missing.

SilverFire replied:
it is not indeed

samdark commented on framework/db/mssql/QueryBuilder.php:
```diff
@@ -46,16 +46,15 @@ class QueryBuilder extends \yii\db\QueryBuilder
-        '\\' => '[\\]',
-    ];
-
+    protected function defaultExpressionBuilders()
```
phpdoc missing.

SilverFire replied:
it is not indeed

samdark commented on framework/db/oci/conditions/LikeConditionBuilder.php:
```diff
@@ -0,0 +1,39 @@
+        '!' => '!!',
+    ];
+
+    public function build(ExpressionInterface $condition, &$params = [])
```
phpdoc missing.

samdark commented on framework/db/pgsql/ArrayParser.php:
```diff
@@ -0,0 +1,103 @@
+    /**
+     * @var string Character used in array
+     */
+    protected $delimiter = ',';
```
I think it should be private.

samdark reviewed (approved):
Overall looks solid. Some minor things added as comments.

SilverFire committed:
Added docs, minor code fixes

samdark committed:
Adjusted wording [skip ci]

SilverFire committed:
Merge branch 'master' into pgsql-arrays-expression-v3

 Conflicts:
	framework/UPGRADE.md

SilverFire committed:
Make ColumnSchema to create ArrayExpression object on poppulation

SilverFire committed:
Updated docs

SilverFire committed:
Merge branch 'master' into pgsql-arrays-expression-v3

 Conflicts (PHPDocs mostly):
	framework/UPGRADE.md
	framework/db/cubrid/QueryBuilder.php
	framework/db/mssql/QueryBuilder.php
	framework/db/oci/QueryBuilder.php
	framework/db/sqlite/QueryBuilder.php

SilverFire committed:
Fixed logical error in ConjunctionConditionBuilder

SilverFire committed:
Fixed one more logical error in ConjunctionConditionBuilder

samdark commented on framework/UPGRADE.md:
```diff
@@ -58,6 +58,14 @@ Upgrade from Yii 2.0.13
+  `instanceof yii\db\Expression` in your code, you might consider changing that to checking for the interface and use the newly
+  introduced methods to retrieve the expression content.
+
+* TODO: `['value', \PDO::PARAM_STR]` syntax deprecation.
```
Should be done before merge.

samdark commented on framework/UPGRADE.md:
```diff
@@ -58,6 +58,14 @@ Upgrade from Yii 2.0.13
+
+* TODO: `['value', \PDO::PARAM_STR]` syntax deprecation.
+
+* TODO: `yii\db\QueryBuilder::conditionBuilders` is no longer used.
```
Should be done before merge.

SilverFire committed:
Updated docs

samdark committed:
Fixed typo [skip ci]

samdark committed:
Adjusted changelog to reflect changes better [skip ci]

samdark commented:
Merged! That is a huge improvement over query builder architecture and I'm glad that Yii has it now. Will allow to easily extend QB syntax in any direction possible. Thank you very much for working on it and finishing it.

qwerty199369 commented:
Excellent! Can't wait to use it! 👍

Blizzke commented:
Was there ever a deprecation period regarding using arrays for bindValues? 
Because the switch to `PdoValue` without having some kind of grace period broke `yii2-audit`.

SilverFire commented:
Yes, it is deprecated since 2.0.14 and will be completely removed in 2.1.
Until that time arrays are silently converted to `PdoValue` objects.

Blizzke commented:
I assume that you are referring to regular inserts only then?

I have a [bug report](https://github.com/bedezign/yii2-audit/issues/238) on the yii2-audit component. It stopped working since 2.0.14. 
We traced it back to the changes in `bindValues()`.

Our component directly calls bindValues() for batch insertion (it is a public function after all). 
The call [can be found here](https://github.com/bedezign/yii2-audit/blob/master/src/models/AuditEntry.php#L127) and has worked for the last 2 years but fails since this PR.

As it is now, there is no grace period for code directly calling `bindValues()` since the `is_array` part was completely replaced:

<img width="981" alt="screen shot 2018-03-02 at 09 27 45" src="https://user-images.githubusercontent.com/231572/36889846-05ef91e8-1dfc-11e8-96bc-8f70964176b1.png">

SilverFire commented:
This break was not intended. Could you create a separate issue? We can probably fix it in 2.0.14.2
